weaviate/batch/crud_batch.py

```diff
@@ -1517,10 +1598,79 @@ def connection_error_retries(self) -> int:
 
     @connection_error_retries.setter
     def connection_error_retries(self, value: int) -> None:
-
         _check_non_negative(value, "connection_error_retries", int)
         self._connection_error_retries = value
 
+    def _retry_on_error(self, response: List[Dict[str, Any]], data_type: str) -> Batch:
+        if data_type == "objects":
+            return self._retry_failed_objects(response)
+        else:
+            return self._retry_failed_references(response)
+
+    def _retry_failed_objects(self, response: List[Dict[str, Any]]) -> Batch:
+        new_batch = ObjectsBatchRequest()
+        for obj in response:
+            if self._skip_objects_retry(obj):
+                continue
+
+            new_batch.add(
+                class_name=obj["class"],
+                data_object=obj["properties"],
+                uuid=obj["id"],
+                vector=obj.get("vector", None),
+            )
+        return new_batch
+
+    def _retry_failed_references(self, response: List[Dict[str, Any]]) -> Batch:
```

Comment:
Good idea, is the order in the response guaranteed? Eg if we send [obj1, obj2, obj3], is the response in the same order in all cases?

Reply:
Hmmm, that is a good question. I do not know this.

Reply:
If we don't know if the order of the responses is always the same, we can't change it, right?

Reply:
This is how the response actually looks like:
```
[{'from': 'weaviate://localhost/Test/8699f59c-894f-466a-83c5-1c226fa2d3a6/crossRef',
  'to': 'weaviate://localhost/Test/aba749c3-e430-4003-9499-e859427207a6',
  'result': {'status': 'SUCCESS'}},
 {'from': 'weaviate://localhost/Test/aba749c3-e430-4003-9499-e859427207a6/crossRef',
  'to': 'weaviate://localhost/Test/8699f59c-894f-466a-83c5-1c226fa2d3a6',
  'result': {'status': 'SUCCESS'}}]
```
and the payload is just a list of `dict`s with the key-value for `from` and `to`. So we do not need the index we extract these from the response.

And here is the object creation response:
```
[{'class': 'Test',
  'creationTimeUnix': 1674141010341,
  'id': '74169e59-3405-47be-9dcc-4202e1fa7b15',
  'lastUpdateTimeUnix': 1674141010341,
  'properties': {'test': 1, 'test2': 2},
  'vector': [1, 2, 3],
  'deprecations': None,
  'result': {}},
 {'class': 'Test',
  'creationTimeUnix': 1674141010341,
  'id': '378ca977-672c-4671-8e59-e68b8ebc882c',
  'lastUpdateTimeUnix': 1674141010341,
  'properties': {'test': 1, 'test2': 2},
  'vector': [1, 2, 3],
  'deprecations': None,
  'result': {}},
 {'class': 'Test',
  'creationTimeUnix': 1674141010341,
  'id': '4871a74d-cf3b-4d39-94c0-86f0963174f2',
  'lastUpdateTimeUnix': 1674141010341,
  'properties': {'test': 1, 'test2': 2},
  'vector': [1, 2, 3],
  'deprecations': None,
  'result': {}}]
```

I think we can add a protected method to each `BatchRequest` (like _add_from_response(respons_item: Dict[str, Any])), where the reference one gets only the `to`/`from` and the objects one extracts `class`/`properties`/`id` and additionally `vector`, in case one was not provided it will not appear in the response item.

This way both `BatchRequests` have same APIs and we do not need to do any parsing but just pass the response item to the empty `BatchRequest`

BTW Weaviate always returns the object back, here is an example for objects:
```
[{'class': 'Test',
  'creationTimeUnix': 1674141346320,
  'id': 'c8fe66c5-76ff-4fdf-a226-05bcd8b8487a',
  'lastUpdateTimeUnix': 1674141346320,
  'properties': {'test': 1, 'test2': '2'},
  'vector': [1, 2, 3],
  'deprecations': None,
  'result': {'errors': {'error': [{'message': "invalid number property 'test2' on class 'Test': requires a float, the given value is '2'"}]}}},
 {'class': 'Test',
  'creationTimeUnix': 1674141346320,
  'id': 'a4e0c869-0714-405a-a72d-c004dc83cbf9',
  'lastUpdateTimeUnix': 1674141346320,
  'properties': {'test': 1, 'test2': '2'},
  'vector': [1, 2, 3],
  'deprecations': None,
  'result': {'errors': {'error': [{'message': "invalid number property 'test2' on class 'Test': requires a float, the given value is '2'"}]}}}]
```

What do you think?

Reply:
please have a look!